server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorDriver.java

```diff
@@ -456,63 +518,93 @@ public void moveSegmentOut(final String sequenceName, final List<SegmentIdentifi
    * internally pushes the segments to the deep storage first, and then publishes the metadata to the metadata storage.
    *
-   * @param publisher segment publisher
-   * @param committer committer
+   * @param publisher     segment publisher
+   * @param committer     committer
    * @param sequenceNames a collection of sequence names to be published
    *
    * @return a {@link ListenableFuture} for the submitted task which removes published {@code sequenceNames} from
-   *         {@code activeSegments} and {@code publishPendingSegments}
+   * {@code activeSegments} and {@code publishPendingSegments}
    */
   public ListenableFuture<SegmentsAndMetadata> publish(
       final TransactionalSegmentPublisher publisher,
       final Committer committer,
       final Collection<String> sequenceNames
   )
   {
-    final List<SegmentIdentifier> theSegments;
-    synchronized (activeSegments) {
-      theSegments = sequenceNames.stream()
-                                 .map(publishPendingSegments::remove)
-                                 .filter(Objects::nonNull)
-                                 .flatMap(Collection::stream)
-                                 .collect(Collectors.toList());
-      sequenceNames.forEach(activeSegments::remove);
+    final List<SegmentIdentifier> theSegments = new ArrayList<>();
+    synchronized (segments) {
+      sequenceNames.stream()
+                   .filter(sequenceName -> !publishingSequences.contains(sequenceName))
+                   .forEach(sequenceName -> {
+                     if (segments.containsKey(sequenceName)) {
+                       segments.get(sequenceName)
+                               .values()
+                               .stream()
+                               .flatMap(Collection::stream)
+                               .forEach(segmentWithState -> {
+                                 segmentWithState.state = SegmentState.PUBLISHING;
+                                 theSegments.add(segmentWithState.getSegmentIdentifier());
+                               });
+                     }
+                   });
     }
 
-    return publish(publisher, wrapCommitter(committer), theSegments);
+    final ListenableFuture<SegmentsAndMetadata> publishFuture = publish(
+        publisher,
+        wrapCommitter(committer),
+        theSegments
+    );
+
+    Futures.addCallback(
+        publishFuture,
+        new FutureCallback<SegmentsAndMetadata>()
+        {
+          @Override
+          public void onSuccess(SegmentsAndMetadata result)
+          {
+            if (result != null) {
+              publishingSequences.removeAll(sequenceNames);
+              sequenceNames.forEach(segments::remove);
+            }
+          }
+
+          @Override
+          public void onFailure(Throwable t)
+          {
+            // Do nothing, caller should handle the exception
+            log.error("Error publishing sequences [%s]", sequenceNames);
```

Comment:
KafkaIndexTask actually fails when there is any exception during publish and also logs that exception. So, just wanted to prevent duplicate traces, I guess I can add it here as well, doesn't hurt.

Reply:
For both `onSuccess` and `onFailure` I am not logging much and relying on caller to log and take appropriate action. Do you think it is ok, its same as the current behavior ?

Reply:
Got it, that seems fine to me.